### DISCUSSION TITLE
Shifted Dirichlet boundary condition and associated resolution

### DISCUSSION THREAD

**Discussion by pmguilcher** [mfem/mfem#2253] [Q&A]
Hello,

I will try to explain as much as possible my problem, which is about imposing a Dirichlet boundary condition **inside** an element. It is in fact related to embedded boundaries, for instance with Level Set.
Let's say we have a triangle element as follows:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;       o    node 1
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /  &nbsp;\\
&nbsp;&nbsp;&nbsp;    /  x  &nbsp;&nbsp;&nbsp;\\          &nbsp;&nbsp; &nbsp;&nbsp;                           x is a particular point in the element
&nbsp;  / &nbsp; &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\\
o ---------o    node 3
node 2

I want to add a constraint on the numerical solution _u_, for instance I want _u(x) = u_imposed_, and solve for _u_ directly with this constraint.
One possibility is to write the FEM approximation of u(x), as a linear combination of the solution u1, u2 and u3 on nodes 1, 2 and 3. I will get something in this way:

_a1.u1 + a2.u2 + a3.u3 = u_imposed_

with _a1_, _a2_ and _a3_ known coefficients related to the FEM approximation.

If we consider that node 1 is a boundary node with a Dirichlet BC, we can just remove the corresponding row  in the LHS of the linear system, and corresponding row in the RHS, that were previously computed to form the linear system. And then use the previous equation (linear combination of u1, u2 and u3) instead.
But the matrix is no longer symmetric.
I was wondering if it is complex to implement, and in which part (I guess I should modify the way to impose the essential bc). And can I use any solver then? Or should I use a specific solver dedicated to non-symmetric matrices?

Thank you
Pierre

**Comment by vladotomov** [comment]
Hi Pierre,

You can find examples for shifted Dirichlet BC (weak enforcement) here #2043.
You can find examples for constraint-type BC (strong enforcement) here #1842.

Seems that you want to do some mix between the above two that I don't quite understand. Is `x` the position of your zero level set? Will the first link above work for you? And yes, for that we use a non-symmetric solver.

Vladimir

**Comment by pmguilcher** [comment]
Hi Vladimir,

Thank you for your quick answer, which is useful.

You're right, my idea is to perform some mix between the shifted Dirichlet BC, and the constraint-type BC.

In the shifted Dirichlet BC, the true boundary is outside of the computational domain, which means we need to perform some back-propagation of the BC from that boundary, to the numerical boundary. To keep a good spatial order, a Taylor expansion is required, with estimation of the derivatives and so on. 
But if the true boundary is _inside_ the computational domain, then we don't need any Taylor expansion, but just use the basis functions to estimate our numerical estimation against the imposed bc. It should be must simpler?

Regarding the constraint-type BC method, the underlying algorithm is a non-linear one? Like Newton or Broyden?

Edit: you're right, the x position is the position of the Level Set, where we want to apply a Dirichlet condition.

Pierre

**Reply by vladotomov** [reply]
The constrained solvers in #1842 are for linear systems.

Yes the idea sounds simpler. For DG spaces you can probably do it element-wise. But how are you planning to choose the exact points (`x`) where you will enforce these "inside basis estimation" conditions? Are you assuming the zero level set curve is straight inside each cell? Also, seems at some stage this will involve an inverse mapping between physical and reference positions.

**Reply by pmguilcher** [reply]
In fact, on the hanging nodes (outside of the **true** computational domain), you already know the level set function, and also its derivative. So you can get  an approximation of the zero levet set curve position, attached to each hanging node. And then you need to determine the element index where is the zero let set point.
